monai/metrics/metric.py

```diff
@@ -0,0 +1,199 @@
+
+        Args:
+            y_pred: the model prediction data to compute, must be a list of `channel-first` Tensor
+                or a `batch-first` Tensor.
+            y: the ground truth to compute, must be a list of `channel-first` Tensor
+                or a `batch-first` Tensor.
+
+        """
+        ret: TensorOrList
+        if isinstance(y_pred, (list, tuple)) or isinstance(y, (list, tuple)):
+            # if y_pred or y is a list of channel-first data, add batch dim and compute metric
+            ret_: List[torch.Tensor] = self._compute_list(y_pred, y)
+            # concat the list of results
+            if isinstance(ret_[0], torch.Tensor):
+                ret = torch.cat(ret_, dim=0)
+            elif isinstance(ret_[0], (list, tuple)) and all([isinstance(i, torch.Tensor) for i in ret_[0]]):
+                # if _compute_tensor() returned not only 1 Tensor, concat them separately
+                ret = [torch.cat([k[i] for k in ret_], dim=0) for i in range(len(ret_[0]))]
+            else:
+                # if not expected data type, return raw results directly
+                ret = ret_
+        elif isinstance(y_pred, torch.Tensor):
+            y_ = y.detach() if y is not None and isinstance(y, torch.Tensor) else None
+            ret = self._compute_tensor(y_pred.detach(), y_)
+        else:
+            raise ValueError("y_pred or y must be a list of `channel-first` Tensors or a `batch-first` Tensor.")
+
+        return ret
+
+    def _compute_list(self, y_pred: TensorOrList, y: Optional[TensorOrList] = None):
+        """
+        Excute the computation for every item of a list.
+        Subclass may enhance the operation with multi-threads to accelerate.
+
+        """
+        if y is not None:
+            return [
+                self._compute_tensor(p_.detach().unsqueeze(0), y_.detach().unsqueeze(0)) for p_, y_ in zip(y_pred, y)
+            ]
+        else:
+            return [self._compute_tensor(p_.detach().unsqueeze(0), None) for p_ in y_pred]
+
+    @abstractmethod
+    def _compute_tensor(self, y_pred: torch.Tensor, y: Optional[torch.Tensor] = None):
+        """
+        computation logic for tensors, input data should be `batch-first` Tensor.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+
+class Cumulative:
+    """
+    Utility class for the typical cumulative computation process based on PyTorch Tensor data.
+    Execute the steps referring to below examples::
+
+        cum = Cumulative(buffer_count=2)
+        cum.add(x, y)
+        cum.add(a, b)
+        cum.add(c, d)
+        cum.agrregate()
+        result = cum.get_synced_tensors()
+        cum.reset()
+
+    Args:
+        buffer_num: the number of buffers to create to cumulate PyTorch Tensors,
+            usually every Tensor should map to a buffer.
+
+    """
+
+    def __init__(self, buffer_num: int = 1):
+        self.buffer_num = buffer_num
```

Comment:
instead of defining `buffer_num`, it could be automatically determined at the first time `self.add`?

Reply:
Sounds like a good idea, let me have a try soon.
Thanks.

Reply:
Changed to automatically detect buffer size in the latest commit.
Thanks for the suggestion.

Thanks.